### PR TITLE
Rewrite frame state to use a struct for shapes, splitting scalar and size, prep for stride

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100815
* __->__ #101250

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -165,6 +165,12 @@ def erase(self):
         self._example = None
 
 
+@dataclasses.dataclass
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:168]: Frozen or not?

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -885,18 +891,25 @@ def wrap_unspecialized_primitive(self, value):
...
+                    frame_state_entry = FrameStateSizeEntry(scalar=value, size=None)
                 else:
-                    curr_size = self.tx.output.frame_state[name]
-                    if curr_size != value:
-                        curr_size = None
-                self.tx.output.frame_state[name] = curr_size
+                    frame_state_entry = self.tx.output.frame_state[name]
+                    if frame_state_entry.scalar != value:
+                        frame_state_entry.scalar = None
+                self.tx.output.frame_state[name] = frame_state_entry
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:903]: If the struct was already set you technically didn't need this

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Test would be good

**Comment by voznesenskym** [comment]
> Test would be good

Should be covered entirely by unspec tests, and test/dynamo/test_recompiles.py `RecompileTests`

**Comment by voznesenskym** [comment]
> > Test would be good
> 
> Should be covered entirely by unspec tests, and test/dynamo/test_recompiles.py `RecompileTests`

Oh well actually I think I can add a test that does a swap on an fn changing from int to tensor and back.

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:903]: Yes, but I like the readability this offers.

**Pull request closed without merging**